python/lsst/analysis/drp/gatherResourceStatistics.py

```diff
@@ -0,0 +1,119 @@
+# This file is part of analysis_drp.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from collections import defaultdict
+
+__all__ = ()
+
+from lsst.pex.config import ListField
+from lsst.pipe.base import PipelineTaskConnections, PipelineTask, PipelineTaskConfig
+from lsst.pipe.base import connectionTypes as ct
+
+
+class GatherResourceStatisticsConnections(PipelineTaskConnections, dimensions=()):
+    output_table = ct.Output(
+        "resource_statistics",
+        storageClass="DataFrame",
+        dimensions=(),
+        doc=(
+            "Table that aggregates memory and CPU usage statistics from one "
+            "or more tasks. "
+            "This will have one row for each data ID, with columns for each "
+            "task or method's memory usage and runtime."
+        )
+    )
+
+    def __init__(self, *, config):
+        super().__init__(config=config)
+        # Inject one new input connection for each configured label.
+        for label in config.grouped_labels():
+            self.inputs.add(label)
+            dataset_type_name = f"{label}_metadata"
+            self.allConnections[label] = ct.Input(
+                dataset_type_name,
+                storageClass="TaskMetadata",
+                dimensions=config.dimensions,
+                multiple=True,
+                deferLoad=True,
+            )
+            # No actual name overrides permitted (these connections aren't
+            # individually configurable, because they're dynamically added),
+            # but for now we need to populate these internal things manually
+            # because super().__init__ can only populate them with the usual
+            # static connections.
+            self._nameOverrides[label] = dataset_type_name
+            self._typeNameToVarName[dataset_type_name] = label
```

Comment:
@natelust , here's a first draft of an attempt to write a connections class with some dynamic inputs, driven by configuration.  I think the two lines above are the worse bits, but it'd certainly also be nice to have a clear way to set both `self.inputs` and `self.allConnections` at the same time consistently, and to make it safe to do that either before or after delegating to `super`.

Curious to get your thoughts on both what I've done given the constraints we have now, and how we might make it easier (but feel free to just go ahead and work on the latter if you have a clear idea of what you want to do).